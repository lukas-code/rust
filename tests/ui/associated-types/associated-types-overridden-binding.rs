//@revisions: pass fail
//@[pass] check-pass

#![feature(trait_alias)]

trait Foo: Iterator<Item = i32> {}
trait Bar: Foo<Item = u32> {}

trait I32Iterator = Iterator<Item = i32>;
trait U32Iterator = I32Iterator<Item = u32>;

#[cfg(fail)]
fn normalize_bar<T: Bar>(_: T::Item) {}
//[fail]~^ ERROR type annotations needed
//[fail]~| ERROR type annotations needed

#[cfg(fail)]
fn normalize_u32iterator<T: U32Iterator>(_: T::Item) {}
//[fail]~^ ERROR type annotations needed
//[fail]~| ERROR type annotations needed

fn main() {
    let _: &dyn Foo<Item = u32>; // OK
    let _: &dyn I32Iterator<Item = u32>; // OK
}
