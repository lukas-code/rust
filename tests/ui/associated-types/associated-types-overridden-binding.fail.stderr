error[E0284]: type annotations needed
  --> $DIR/associated-types-overridden-binding.rs:13:29
   |
LL | fn normalize_bar<T: Bar>(_: T::Item) {}
   |                             ^^^^^^^ cannot infer type
   |
   = note: cannot satisfy `<T as Iterator>::Item == _`

error[E0284]: type annotations needed
  --> $DIR/associated-types-overridden-binding.rs:18:45
   |
LL | fn normalize_u32iterator<T: U32Iterator>(_: T::Item) {}
   |                                             ^^^^^^^ cannot infer type
   |
   = note: cannot satisfy `<T as Iterator>::Item == _`

error[E0284]: type annotations needed
  --> $DIR/associated-types-overridden-binding.rs:13:38
   |
LL | fn normalize_bar<T: Bar>(_: T::Item) {}
   |                                      ^^ cannot infer type
   |
   = note: cannot satisfy `<T as Iterator>::Item == _`

error[E0284]: type annotations needed
  --> $DIR/associated-types-overridden-binding.rs:18:54
   |
LL | fn normalize_u32iterator<T: U32Iterator>(_: T::Item) {}
   |                                                      ^^ cannot infer type
   |
   = note: cannot satisfy `<T as Iterator>::Item == _`

error: aborting due to 4 previous errors

For more information about this error, try `rustc --explain E0284`.
